
Import the libraries math (for square root and absolute value) and pandas and numpy for reading csv files.

In [1]:
import math
import pandas as pd
import numpy as np

Make an untagged vector, point, and two tagged vectors, data1 and data2:

In [2]:
point = [1, 0, 0, '?'] #(an unknown tag)
data1 = [1, 1, 1, 'M']
data2 = [1, 3, 0, 'F']

Write code to separate the data (X) from the tag (Y). Your output should be:

The vector [1, 1, 1] has tag M

The vector [1, 3, 0] has tag F

In [3]:
print("The vector " , data1[0:-1], " has tag ", data1[-1])
print("The vector " , data2[0:-1], " has tag ", data2[-1]) #Print the data for vector 2

The vector  [1, 1, 1]  has tag  M
The vector  [1, 3, 0]  has tag  F


Now let's classify the point as either M or F. We'll do this by setting k = 1 and using the Euclidean distance. We'll define that as:

In [4]:
def euclideanDistance(instance1, instance2, length):
   distance = 0
   for x in range(length):
         #print ('x is ' , x)
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance = distance + pow(num1-num2, 2)
   return math.sqrt(distance)

Now, find out the distance between data1 and point and data2 and point. Output what the tag should be using euclideanDistance.

In [5]:
print(euclideanDistance(data1,point,3))

1.4142135623730951


In [6]:
url = 'https://github.com/rosenfa/ai/blob/master/myFile.csv?raw=true'
df=pd.read_csv(url,  header=0, error_bad_lines=False)
#put data in dataset without header line
dataset = np.array(df)

C:\Users\eohay\AppData\Local\Temp\ipykernel_14256\1700079168.py:2: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  df=pd.read_csv(url,  header=0, error_bad_lines=False)


Show that you understand what you read by:

Printing the first two vectors in the file
Printing the Euclidean distance between those two vectors
Here is some code which I think might help get you on your way!

In [7]:
print(len(dataset))
print(euclideanDistance(dataset[0],dataset[1],3))

10
5.744562646538029


Now assume you have a new value for point:

[0,0,100]

How would you classify this vector using the Euclidean distance function given all of the vectors in the file?

In order to help here is a hint:

We suggest defining some type of data structure to store different vectors' distances and their tags like this:

In [8]:
point=[0,0,100]
class distClass:
    dist = -1 #distance of current point from test point
    tag = '-' #tag of current point

You can then add vector distances like this:

In [9]:
def generate_euclidean_distances(point, data,p): #Generate the distances for this point with each vector in the dataset
    distances = [] # list of distances, will hold objects of type distClass
    for vector in data: # Loop through every single vector in database  
        d=euclideanDistance(point,vector,len(vector)-1) # calculate the distance between the vector at database[i] and "point"
        label=vector[-1] # Grab the label of the vector
        if p==0:
            print("The distances between " , point , " and " , vector,  " is " , str(d))
            print(" and the label is " + label)
        obj = distClass() #one record's distance and tag
        obj.dist=d
        obj.tag=label
        distances.append(obj)
    distances.sort(key=lambda x: x.dist)
    return distances

In [10]:
eucDistances=generate_euclidean_distances(point,dataset,0)

The distances between  [0, 0, 100]  and  [0 1 2 'F']  is  98.00510190801293
 and the label is F
The distances between  [0, 0, 100]  and  [1 5 6 'F']  is  94.13819628609845
 and the label is F
The distances between  [0, 0, 100]  and  [1 40 1 'M']  is  106.78014796768171
 and the label is M
The distances between  [0, 0, 100]  and  [2 2 2 'F']  is  98.04080783020915
 and the label is F
The distances between  [0, 0, 100]  and  [0 1 20 'M']  is  80.00624975587844
 and the label is M
The distances between  [0, 0, 100]  and  [1 5 6 'M']  is  94.13819628609845
 and the label is M
The distances between  [0, 0, 100]  and  [1 4 1 'M']  is  99.08582138732059
 and the label is M
The distances between  [0, 0, 100]  and  [2 22 2 'F']  is  100.45894683899488
 and the label is F
The distances between  [0, 0, 100]  and  [2 23 2 'M']  is  100.68266980965493
 and the label is M
The distances between  [0, 0, 100]  and  [2 24 2 'M']  is  100.91580649234291
 and the label is M


Questions:

What is the label for point if k=1?
What is the label for point if k=3?
Would the result be different if we used a different distance function like Hamming or Manhattan?

In [11]:
def determine_tag(distances, k): #This value determines the tag of a point given a list of its distances with another point and
                            # k determines how many neighbors we explore
    tags={}
    max_tag="" #Keep track of the tag that appears the most
    for x in range(k): #Check the k closest neighbors and store how many times we see each tag
        obj=distances[x]
        if obj.tag in tags: #If we have already encountered this tag increment it
            tags[obj.tag]+=1
        else: #else add the tag with a value of one
             tags[obj.tag]=1 
        max_tag = max_tag if tags.get(max_tag, 0) >= tags.get(obj.tag, 0) else obj.tag # check if the new tag is larger than max _tag and store the larger
    return max_tag #return what the point is classified based on its k closest neighbors

In [12]:
print(f"The label for point is {determine_tag(eucDistances,1)} if k=1")
print(f"The label for point is {determine_tag(eucDistances,3)} if k=3")

The label for point is M if k=1
The label for point is M if k=3


In [27]:
#Add code with functions that implement Hamming and Manhattan distances and test
#what the label will be for k=1 and k=3 for all possibilities
#(6 total: 2 Euclidean, 2 Hamming and 2 Manhattan)

def hamming_distance(instance1, instance2, length): #Hamming distance is just checking whether the values are the same and if not increment by 1
   distance = 0
   for x in range(length):
         num1=float(instance1[x])
         num2=float(instance2[x])
         if num1!=num2:
             distance+=1
   return distance

def manhattan_distance(instance1, instance2, length): #Manhattan distance is just summing |num1-num2| accross zipping(instance1,instance2)
    distance = 0
    for x in range(length):
         num1=float(instance1[x])
         num2=float(instance2[x])
         distance+= abs(num1-num2)
    return distance
    

In [30]:
print("Hamming distances")
def generate_hamming_distances(point, data, p): #Generate the distances for this point with each vector in the dataset
    distances = [] # list of distances, will hold objects of type distClass
    for vector in data: # Loop through every single vector in database  
        d=hamming_distance(point,vector,len(vector)-1) # calculate the distance between the vector at database[i] and "point"
        label=vector[-1] # Grab the label of the vector
        if p==0:
            print("The distances between " , point , " and " , vector,  " is " , str(d))
            print(" and the label is " + label)
        obj = distClass() #one record's distance and tag
        obj.dist=d
        obj.tag=label
        distances.append(obj)
    distances.sort(key=lambda x: x.dist) #Sort and return the hamming distance array
    return distances
    
hammingDist= generate_hamming_distances(point, dataset,0)

print("\nManhattan distances")
def generate_mahattan_distances(point, data,p): #Generate the distances for this point with each vector in the dataset
    distances = [] # list of distances, will hold objects of type distClass
    for vector in data: # Loop through every single vector in database  
        d=manhattan_distance(point,vector,len(vector)-1) # calculate the distance between the vector at database[i] and "point"
        label=vector[-1] # Grab the label of the vector
        if p==0:
            print("The distances between " , point , " and " , vector,  " is " , str(d))
            print(" and the label is " + label)
        obj = distClass() #one record's distance and tag
        obj.dist=d
        obj.tag=label
        distances.append(obj)
    distances.sort(key=lambda x: x.dist) #Sort and return the hamming distance array
    return distances
    
manhattanDist= generate_mahattan_distances(point, dataset,0)

Hamming distances
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [0 1 2 'F']  is  2
 and the label is F
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [1 5 6 'F']  is  3
 and the label is F
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [1 40 1 'M']  is  3
 and the label is M
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [2 2 2 'F']  is  3
 and the label is F
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [0 1 20 'M']  is  2
 and the label is M
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [1 5 6 'M']  is  3
 and the label is M
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [1 4 1 'M']  is  3
 and the label is M
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances between  [0, 0, 100]  and  [2 22 2 'F']  is  3
 and the label is F
[0, 0, 100]
[0, 0, 100]
[0, 0, 100]
The distances b

In [26]:
manhattanDist.sort(key=lambda x: x.dist) #Sort the arrays
hammingDist.sort(key=lambda x: x.dist)

In [16]:
print(f"Using hamming distance, the label for point is {determine_tag(hammingDist,1)} if k=1")
print(f"Using hamming distance, the label for point is {determine_tag(hammingDist,3)} if k=3") #With hamming distance it can be either M or F so the program just returns the label that happens to be next

print(f"Using manhattan distance, the label for point is {determine_tag(manhattanDist,1)} if k=1")
print(f"Using manhattan distance, the label for point is {determine_tag(manhattanDist,3)} if k=3")

Using hamming distance, the label for point is F if k=1
Using hamming distance, the label for point is F if k=3
Using manhattan distance, the label for point is M if k=1
Using manhattan distance, the label for point is F if k=3


Up until here is a simplified version of the homework. Below here is the work for the part we will be checking as the basis of your grade:

Now let's look at some bigger files:

mytrain.csv (for training the model)
mytest.csv (for testing)

In [17]:
url = 'https://github.com/rosenfa/ai/blob/master/mytrain.csv?raw=true'
train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
url = 'https://github.com/rosenfa/ai/blob/master/mytest.csv?raw=true'
test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))

C:\Users\eohay\AppData\Local\Temp\ipykernel_14256\3800603745.py:2: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  train_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))
C:\Users\eohay\AppData\Local\Temp\ipykernel_14256\3800603745.py:4: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.


  test_data = np.array(pd.read_csv(url,  header=0, error_bad_lines=False))


In [18]:
print(train_data.shape)# number of records and features
print(train_data)

(100, 30)
[[0.317251 0.540016 0.150514 ... 0.0 0.212007 'F']
 [3.882181 0.0 0.0 ... 0.0 0.0 'F']
 [1.790708 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'F']]


In [19]:
print(test_data.shape)# number of records and features
print(test_data)

(100, 30)
[[2.85985 0.66275 0.0 ... 0.995289 0.0 'F']
 [0.0 12.998726 0.0 ... 0.0 0.0 'F']
 [13.231318 0.0 0.0 ... 0.0 0.0 'F']
 ...
 [0.0 0.0 0.0 ... 0.0 0.0 'M']
 [1.688611 0.0 3.204529 ... 0.0 0.0 'M']
 [0.0 0.0 0.0 ... 0.0 0.0 'M']]


I hope by now you understand where we are going with this :)

Now implement the knn code with 3 different values for k:

k = 1
k = 7
k = 15
and at first use the Euclidean distance. Classify each of the vectors in the test_data dataset using the training data from train_data. Which value for k gave the best accuracy? What accuracy did it give you?

Now see if using Hamming or Manhattan distance give any better results for the same values of k.

Once you are done, you should have a total of 9 different results:

Three accuracy results for the different value of k using the Euclidean Distance
Three accuracy results for the different value of k using the Hamming Distance
Three accuracy results for the different value of k using the Manhattan Distance
Hint: I strongly suggest you base yourself on the code you've seen until this point so you don't have to reinvent the wheel!

In [20]:
# For Euclidean Distance:
euclideanDistances=[] #Data structure to store nearest neighbors for each vector in test data
for test_vector in test_data: # For every vector in the test data generate the euclidean test from it to every vector in the train data
    euclideanDistances.append(generate_euclidean_distances(test_vector, train_data,1))

def generate_classification(test_data_distances, k): # Classify every vector in test_data_distances and return an array representing the classifications 
    classifications=[]
    for distances in test_data_distances: # Loop accross all vectors and classify them based on neighbors
        classifications.append(determine_tag(distances,k))
    return classifications

def determine_accuracy(classifications, test_data): # Determine the accuracy of the model
    correct_classification=[x[-1] for x in test_data] #generate a list corresponding to the correct classification
    score=0 # store how many times we are correct
    
    for guess,correct in zip(classifications,correct_classification): #Check for accuracy by comparing the proposed classification with true classification
        if guess==correct:
            score+=1
    # We can get the accuracy by determining the ratio of correct classification to total classifications
    # We get this by doing # of times correct/ total guesses
    # Where len(classifications) is the total number of classifications and correct is hwo many times we guessed correctly
    return score/len(classifications) * 100
    
 # classification of vectors in test_data where classification[0] is the classification of test_data[0]

classfication_1n=generate_classification(euclideanDistances,1) # classification when k=1
print(f"Using Euclidean Distance when k=1 the model has an accuracy of {determine_accuracy(classfication_1n,test_data)}%")

classfication_7n=generate_classification(euclideanDistances,7) # classification when k=7
print(f"Using Euclidean Distance when k=7 the model has an accuracy of {determine_accuracy(classfication_7n,test_data)}%")

classfication_15n=generate_classification(euclideanDistances,15) # classification when k=15
print(f"Using Euclidean Distance when k=15 the model has an accuracy of {determine_accuracy(classfication_15n,test_data)}%")

print("\nFor Euclidean distance the most accurate was when k=7 with 74.0% which somewhat accurate")
    

Using Euclidean Distance when k=1 the model has an accuracy of 50.0%
Using Euclidean Distance when k=7 the model has an accuracy of 74.0%
Using Euclidean Distance when k=15 the model has an accuracy of 70.0%

For Euclidean distance the most accurate was when k=7 with 74.0% which somewhat accurate


In [21]:
# For Manhattan Distance
manhattanDistances=[] #Data structure to store nearest neighbors for each vector in test data
for test_vector in test_data:
    manhattanDistances.append(generate_mahattan_distances(test_vector,train_data,1))
    
classfication_1n=generate_classification(manhattanDistances,1) # classification when k=1
print(f"Using Manhattan Distance when k=1 the model has an accuracy of {determine_accuracy(classfication_1n,test_data)}%")

classfication_7n=generate_classification(manhattanDistances,7) # classification when k=7
print(f"Using Manhattan Distance when k=7 the model has an accuracy of {determine_accuracy(classfication_7n,test_data)}%")

classfication_15n=generate_classification(manhattanDistances,15) # classification when k=15
print(f"Using Manhattan Distance when k=15 the model has an accuracy of {determine_accuracy(classfication_15n,test_data)}%")

print("\nFor Manhattan distance the most accurate was when k=7 with 69.0% which is not very accurate")
    

Using Manhattan Distance when k=1 the model has an accuracy of 61.0%
Using Manhattan Distance when k=7 the model has an accuracy of 63.0%
Using Manhattan Distance when k=15 the model has an accuracy of 69.0%

For Manhattan distance the most accurate was when k=7 with 69.0% which is not very accurate


In [22]:
# For hamming Distance
hammingDistances=[] #Data structure to store nearest neighbors for each vector in test data
for test_vector in test_data:
    hammingDistances.append(generate_hamming_distances(test_vector, train_data,1))
    
    
classfication_1n=generate_classification(hammingDistances,1) # classification when k=1
print(f"Using Hamming Distance when k=1 the model has an accuracy of {determine_accuracy(classfication_1n,test_data)}%")

classfication_7n=generate_classification(hammingDistances,7) # classification when k=7
print(f"Using Hamming Distance when k=7 the model has an accuracy of {determine_accuracy(classfication_7n,test_data)}%")

classfication_15n=generate_classification(hammingDistances,15) # classification when k=15
print(f"Using Hamming Distance when k=15 the model has an accuracy of {determine_accuracy(classfication_15n,test_data)}%")

print("\nHamming distance was the most accurate when k=1 with 61.0% accuracy")
    

Using Hamming Distance when k=1 the model has an accuracy of 61.0%
Using Hamming Distance when k=7 the model has an accuracy of 55.00000000000001%
Using Hamming Distance when k=15 the model has an accuracy of 56.99999999999999%

Hamming distance was the most accurate when k=1 with 61.0% accuracy


In the end, the most accurate was the Euclidean Distance of k=7 with 74% accuracy. K=7 was the best nearest neighbors as with 1 neighbor the sample set is too small and with k=15 it is too large and could lead to overfitting